### ISSUE TITLE
Make tray icon background color transparent instead of white

### ISSUE THREAD

**Issue by andrewdavidwong** [QubesOS/qubes-issues#2846]
If possible, it would be nice if system tray icons could have transparent backgrounds instead of solid white. This would make them look more like native dom0 tray icons in dark color schemes.

Here's a screenshot of how they currently look, courtesy of user [Epitre](https://groups.google.com/d/topic/qubes-users/SbU_kiPpT4Y/discussion):

![white-tray-icons](https://13366229192823780453.googlegroups.com/attach/9995673d80e02/qubes-example-trayicons.jpg?part=0.1&view=1&vt=ANaJVrGLU7-cWRPjAV8fJreir9gSrb7Ty7UctYO7MEkgwJ8HyshicT5FzInENLD8_ul99jqZ_EoKwFzToRp93lnTnEm_5JEbjO4WX7O7VpZwl3-DvJWV-1c)

(Compare the backgrounds of the the Skype and NetworkManager icons to the Volume and Qubes Manager icons.)

**Comment by marmarek** [comment]
There are different modes for handling icons - see man qubes-guid. None of them offer transparent background, but it could be easily added based on "bg" one.
But first we need to make sure it will be still clearly marked that the icon is from a VM.

**Comment by fepitre** [comment]
I'm working on it. I added the choice of having a background with a specific color for "bg" and "tint". From what I have seen and tested in the code, the "bg" mode seems to not color the background with respect to appvm color? In that case, I could adjust it.

**Comment by fepitre** [comment]
@marmarek Do you know what is the best way to pass a value to vmside from xside? It is intended to pass a color for setting the background in vmside when XCreateSimpleWindow with WhitePixel as background. Without it, the icons look like badly stripped on their border due to the white.

**Comment by marmarek** [comment]
You mean that vmside would fill the background with the color of actual tray background? What if the color changes? What about choosing some different than white (but static) color as a "transparent" one? Like 0xfefefe, or so. Also, in theory each pixel is 32 bits, but in practice only 24 bits are used. Maybe we can have RGBA, instead of RGB there?

**Comment by fepitre** [comment]
Yes the vmside would fill the color of actual tray but if color change, the vm needs to restart...not a so good final way. I will try with a different static color but probably we will have a bad stripping. The best solution would be to use the RGBA indeed. I will investigate.

**Comment by abbluiz** [comment]
Any updates on this? 

Just a suggestion: what if, at least, you could round the corners of the icon or change the white background according to the VM label color? Even though is a small detail, I think dark theme users would benefit a lot from this.

**Comment by marmarek** [comment]
You can set it to VM color in `/etc/qubes/guid.conf`. See `man qubes-guid` for available options.

**Comment by xjeanjean** [comment]
@marmarek This is just completely wrong - this thing doesn't even do anything anymore on 4.1.
The xfce4 theming looks amateurish and I stuck on an eye-hurting light mode just because I could not stand the weird icons with white background for the network manager.

@fepitre could you give us some pointer, even some random half working code ? I'd love to hardcode my preferred color but I have no clue where to start looking. If you already have something hacky able to force a color from the VM, it would be really useful.

```XCreateSimpleWindow with WhitePixel as background.```
Where is that ?

**Comment by fepitre** [comment]
I need to dig the code again but I gave up long time ago for prior tasks. The problem is that it's not something you can customize as is. You need to modify the C code then build the gui-daemon by yourself. This is really not suitable. Anyway, I will try to find the thing you need.

**Comment by dud1337** [comment]
https://github.com/QubesOS/qubes-gui-daemon/blob/master/gui-daemon/trayicon.c
Here is the code. It is quite short and one can see there is some special attention to white colours during the tinting functionality. I am not familiar with plane masks/etc to be able to help.

**Comment by RealityAnomaly** [comment]
<del>I made an attempt at this here: https://github.com/QubesOS/qubes-gui-daemon/pull/167</del>

<del>It still doesn't look quite right (maybe because of tinting weirdness) but it's a lot better than a full-white background IMO.</del>

Decided to rework this so the transparent pixels get passed through from the vmside instead

**Comment by RealityAnomaly** [comment]
OK, so I've done some investigation - while it seems possible in theory to pass 32-bit colour values through by modifying gui-daemon, gui-agent, and the xf86-video-dummy driver, it still won't be capable of drawing the channel unless there is a compositor present, I believe. I don't know if a compositor would need to be present in just the vm or both the vm and host sides, but I am not sure if we even really want to have a compositor, what sort of overhead that would introduce, or the security implications of allowing windows to render with an alpha channel.

I think the most straightforward solution for now would be to do what @fepitre attempted and just set the desired colour value in vmside.c in the place of WhitePixel being passed into XCreateWindow; this obviously has the limitation of not being able to be updated on the fly, but it does solve the stripping problem. Also this has the side effect of the background being tinted too.

**Comment by RealityAnomaly** [comment]
https://github.com/QubesOS/qubes-gui-agent-linux/pull/238 (along with its associated patch, https://github.com/QubesOS/qubes-core-admin/pull/699) should mitigate this to an extent. The background still does get tinted, but there's no reliable way around that without an alpha channel.

<img width="76" height="30" alt="Image" src="https://github.com/user-attachments/assets/c6cf089b-4fa5-4429-9a3c-093ff4f49032" />

**Comment by marmarek** [comment]
So, is the idea to have a configurable background color on the VM side, that matches panel background? But as you noticed, it will be tinted anyway (for some colors+icons maybe not a big issue, but for others might be). What was wrong with that earlier approach of painting it at the gui daemon side (based on corner pixel color or something similar)? Might need a bit smarter approach, to not paint white pixels within the icon too... Maybe set the background to some less common color, like the 0xfefefe (or 0xfffffe?) used somewhere already?  In other words - I would like to avoid giving VM more information about your environment in dom0...

As for the compositor - we do enable compositing in VM (otherwise we wouldn't have per-window buffers available). And at least in Xfce we do that in dom0 too. But I'm not sure about other window managers or how that would work with sys-gui... That said, I'm not sure if compositor is really required for that. After all, non-vm icons don't have this solid rectangle issue.
Anyway, even if dom0 wouldn't support full 32-bit values, having VM sending that info would be useful already - this way, you can color it in GUI daemon more accurately (we do manual re-coloring anyway, so it wouldn't complicate things much). So, if you know how to make VM use full 32-bit value for icons, I'd go this way. Otherwise, simply use a bit less common background that can be re-colored on the gui-daemon side (ideally - to transparent, but if not, then to the configured color).

**Comment by RealityAnomaly** [comment]
@marmarek One of my concerns with using a defined less-common colour is that I think it would break some icons rather unexpectedly? Sure, it'd probably work in most cases, but you might see some icons that happen to contain that colour with random streaks as it gets replaced.

I also had a problem with my earlier approach, which is that the icon actually blends with its window's background (as defined in vmside.c). This means that any pixel replacement on the dom0 side will end up making the icon look weird, as the replacement happens after blending with the background. I've illustrated the problem here, on an Xfce light vs dark theme:

<img width="70" height="150" alt="Image" src="https://github.com/user-attachments/assets/2c04c621-0f19-42e1-9c16-7ea8e9464d4b" />

The borders around the monitors are actually different colours. Bottom image is the top but with its BG replaced with the BG of the middle - it looks wrong. Say you made the background chroma key green - you wouldn't be able to just replace that colour on the dom0 side, because the entire icon gets blended against that green, not just its background.

**Comment by Joeviocoe** [comment]
Is this still on the roadmap?

Or does guid.conf > trayicon_mode do anything anymore?  It says deprecated and to use qvm-features, but that man page doesn't mention tray icons.
